### PR TITLE
Fix crash when deleting many moving dynamic objects.

### PR DESCRIPTION
**Pull request by AndrewMeadows**
- fix a recently added crash mode: when deleting lots of dynamic objects at once

### CODE REVIEW

File: libraries/physics/src/PhysicalEntitySimulation.cpp [hunk 1]
```diff
@@ -169,6 +169,7 @@ void PhysicalEntitySimulation::getObjectsToRemoveFromPhysics(VectorOfMotionState
         EntityMotionState* motionState = static_cast<EntityMotionState*>(entity->getPhysicsInfo());
         assert(motionState);
         _pendingChanges.remove(motionState);
+        _outgoingChanges.remove(motionState);
```
Comment (AndrewMeadows) [inline libraries/physics/src/PhysicalEntitySimulation.cpp]: This is the actual fix: don't forget to remove dangling pointers from **PhysicalEntitySimulation::_outgoingChanges** list.

The rest of the changes are just cleanup that I noticed while trying to track this down.

The reason this crash mode is new is that we stopped checking for invalid pointers in **EntityMotionState::isCandidateForOwnership()**, which used to trigger a cleanup of the pointer inside **_outgoingChanges**.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 84fb983da763d237e1b01323d5935b8597df240d available at: https://gist.github.com/f34e46717aeebfe2d10e<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7061.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7061.dmg

**Comment by OmegaHeron** [comment]
@AndrewMeadows do you have a copy of script we used to test this before (mass create/delete entities)?  I didn't keep a ref copy and can't find the old PR on this.

**Comment by AndrewMeadows** [comment]
The internal issue is here:  https://app.asana.com/0/26623005675174/89386472074953

To reproduce: run the following script which creates a pile of dynamic boxes with 10sec lifetimes.  The interface should crash (or not) when the pile is deleted

    "use strict";
    /*jslint nomen: true, plusplus: true, vars: true*/
    var Vec3, Quat, MyAvatar, Entities, Camera, Script, print;
    var NUMBER_TO_CREATE = 200;
    var LIFETIME = 10; // seconds
    var SIZEX = 0.5;
    var SIZEY = 1.5;
    var SIZEZ = 0.5;
    var TYPE = "Box";
    var RATE_PER_SECOND = 600;
    var SCRIPT_INTERVAL = 100;
    var GRAVITY = { x: 0, y: -9.8, z: 0 };
    var VELOCITY = { x: 0.0, y: 0, z: 0 };
    var ANGULAR_VELOCITY = { x: 1, y: 1, z: 1 };
    var DAMPING = 0.5;
    var ANGULAR_DAMPING = 0.5;
    var RANGE = 1;
    var HOW_FAR_IN_FRONT_OF_ME = 9;
    var HOW_FAR_UP = 3;
    //
    var totalCreated = 0;
    var offset = Vec3.sum(Vec3.multiply(HOW_FAR_UP, Vec3.UNIT_Y),
                          Vec3.multiply(HOW_FAR_IN_FRONT_OF_ME, Quat.getFront(Camera.orientation)));
    var center = Vec3.sum(MyAvatar.position, offset);
    function randomVector(range) {
        return {
            x: (Math.random() -  0.5) * range.x,
            y: (Math.random() -  0.5) * range.y,
            z: (Math.random() -  0.5) * range.z
        };
    }
    //
    Script.setInterval(function () {
        if (!Entities.serversExist() || !Entities.canRez()) {
            return;
        }
        if (totalCreated >= NUMBER_TO_CREATE) {
            print("Created " + totalCreated + " tribbles.");
            Script.stop();
        }
        var i, numToCreate = RATE_PER_SECOND * (SCRIPT_INTERVAL / 1000.0);
        for (i = 0; (i < numToCreate) && (totalCreated < NUMBER_TO_CREATE); i++) {
            Entities.addEntity({
                type: TYPE,
                name: "tribble-" + totalCreated,
                position: Vec3.sum(center, randomVector({ x: RANGE, y: RANGE, z: RANGE })),
                dimensions: {x: SIZEX, y: SIZEY, z: SIZEZ},
                color: {red: Math.random() * 255, green: Math.random() * 255, blue: Math.random() * 255},
                velocity: VELOCITY,
                angularVelocity: Vec3.multiply(Math.random(), ANGULAR_VELOCITY),
                damping: DAMPING,
                angularDamping: ANGULAR_DAMPING,
                gravity: GRAVITY,
                dynamic: true,
                lifetime: LIFETIME
            });
            totalCreated++;
        }
    }, SCRIPT_INTERVAL);

**Comment by OmegaHeron** [comment]
Tested Windows 7 against production build 4158.

Note - I couldn't get 4158 to crash with this - ran script 10 times without issue.  Ran with this PR 15 cycles, no crash.  Only thing noted was some objects created didn't delete (seen on 2 of 15 passes).

**Comment by OmegaHeron** [comment]
@AndrewMeadows 

Tested again - was able to crash production build 4158 after 3 cycles, but, also crashed this PR after 5 - stack trace follows; (This is for _this_ PR)

>	interface.exe!PhysicsEngine::bump(ObjectMotionState * motionState) Line 410	C++
 	interface.exe!PhysicsEngine::removeObjects(const QVector<ObjectMotionState *> & objects) Line 152	C++
 	interface.exe!Application::update(float deltaTime) Line 3124	C++
 	interface.exe!Application::idle(unsigned __int64 now) Line 2454	C++
 	interface.exe!Application::paintGL() Line 1332	C++
 	interface.exe!Application::event(QEvent * event) Line 1716	C++
 	Qt5Widgets.dll!0000000064c95b96()	Unknown
 	Qt5Widgets.dll!0000000064c94595()	Unknown
 	Qt5Core.dll!0000000066b5c4e8()	Unknown
 	Qt5Core.dll!0000000066b5e144()	Unknown
 	qwindows.dll!000007fee5fa096f()	Unknown
 	Qt5Core.dll!0000000066ba1ccc()	Unknown
 	[External Code]	
 	Qt5Core.dll!0000000066ba1389()	Unknown
 	qwindows.dll!000007fee5fa089e()	Unknown
 	Qt5Core.dll!0000000066b59191()	Unknown
 	Qt5Core.dll!0000000066b5b667()	Unknown
 	interface.exe!main(int argc, const char * * argv) Line 136	C++
 	[External Code]	

The thread 0x2158 has exited with code 0 (0x0).
The thread 0x2aa8 has exited with code 0 (0x0).
The thread 0x23a4 has exited with code 0 (0x0).
The thread 0x1f70 has exited with code 0 (0x0).
'interface.exe' (Win32): Unloaded 'C:\WINDOWS\System32\wpdshext.dll'
Unhandled exception at 0x000000013FD8A2FE in interface.exe: 0xC0000005: Access violation reading location 0x0000000000000018.

The program '[656] interface.exe' has exited with code 0 (0x0).

**Comment by OmegaHeron** [comment]
And again, after 12 passes

>	interface.exe!QHash<EntityItemID,std::shared_ptr<EntityTreeElement> >::value(const EntityItemID & akey) Line 658	C++
 	interface.exe!EntityTree::getContainingElement(const EntityItemID & entityItemID) Line 1188	C++
 	interface.exe!EntityTree::findEntityByEntityItemID(const EntityItemID & entityID) Line 670	C++
 	interface.exe!EntityTreeRenderer::isCollisionOwner(const QUuid & myNodeID, std::shared_ptr<EntityTree> entityTree, const EntityItemID & id, const Collision & collision) Line 703	C++
 	interface.exe!EntityTreeRenderer::playEntityCollisionSound(const QUuid & myNodeID, std::shared_ptr<EntityTree> entityTree, const EntityItemID & id, const Collision & collision) Line 729	C++
 	interface.exe!EntityTreeRenderer::entityCollisionWithEntity(const EntityItemID & idA, const EntityItemID & idB, const Collision & collision) Line 792	C++
 	Qt5Core.dll!00000000664cb287()	Unknown
 	interface.exe!EntitySimulation::entityCollisionWithEntity(const EntityItemID & _t1, const EntityItemID & _t2, const Collision & _t3) Line 146	C++
 	interface.exe!PhysicalEntitySimulation::handleCollisionEvents(const QVector<Collision> & collisionEvents) Line 297	C++
 	interface.exe!Application::update(float deltaTime) Line 3178	C++
 	interface.exe!Application::idle(unsigned __int64 now) Line 2454	C++
 	interface.exe!Application::paintGL() Line 1332	C++
 	interface.exe!Application::event(QEvent * event) Line 1716	C++
 	Qt5Widgets.dll!00000000642c5b96()	Unknown
 	Qt5Widgets.dll!00000000642c4595()	Unknown
 	Qt5Core.dll!00000000664ac4e8()	Unknown
 	Qt5Core.dll!00000000664ae144()	Unknown
 	qwindows.dll!000007fee428096f()	Unknown
 	Qt5Core.dll!00000000664f1ccc()	Unknown
 	[External Code]	
 	Qt5Core.dll!00000000664f1389()	Unknown
 	qwindows.dll!000007fee428089e()	Unknown
 	Qt5Core.dll!00000000664a9191()	Unknown
 	Qt5Core.dll!00000000664ab667()	Unknown
 	interface.exe!main(int argc, const char * * argv) Line 136	C++
 	[External Code]	

The thread 0xf84 has exited with code 0 (0x0).
The thread 0x2894 has exited with code 0 (0x0).
The thread 0x850 has exited with code 0 (0x0).
Unhandled exception at 0x000000013F645873 in interface.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.

The program '[10728] interface.exe' has exited with code 0 (0x0).

**Comment by AndrewMeadows** [comment]
@OmegaHeron: I was trying to reproduce your first crash mode and kept hitting this one instead (on linux, in debug mode -- was hitting an assert, so I guess it might not crash for a release build).

I think I have a fix for your first crash mode.  I'll be putting that together later today.

The second crash mode... did that happen when the objects were deleted?  or before their timer had expired?

**Comment by OmegaHeron** [comment]
Second one seems to have been before delete.

**Pull request merged**